From 06de0ed2b191070c5048f7e7e7b0e190d1f35fbf Mon Sep 17 00:00:00 2001
From: Wim Taymans <wtaymans@redhat.com>
Date: Mon, 7 Sep 2026 19:28:44 +0200
Subject: [PATCH] metadata: remove pending pong on unbind

A client that binds and unbinds will leave the pending counter
incremented forever and leave old clients without updates.

Fix this by making sure that we clear any pending pong when the client
unbinds before we could ping/pong.

Fixes #5445

diff --git a/src/modules/module-metadata/metadata.c b/src/modules/module-metadata/metadata.c
index c0aa37110..39f07beda 100644
--- a/src/modules/module-metadata/metadata.c
+++ b/src/modules/module-metadata/metadata.c
@@ -102,10 +102,19 @@ static const struct pw_metadata_methods metadata_methods = {
 	.clear = metadata_clear,
 };
 
+static void remove_pending(struct resource_data *d)
+{
+	if (d->pong_seq != 0) {
+		pw_impl_client_set_busy(pw_resource_get_client(d->resource), false);
+		d->pong_seq = 0;
+		d->impl->pending--;
+	}
+}
 
 static void global_unbind(void *data)
 {
 	struct resource_data *d = data;
+	remove_pending(d);
 	if (d->resource) {
 	        spa_hook_remove(&d->resource_listener);
 	        spa_hook_remove(&d->object_listener);
@@ -119,15 +128,6 @@ static const struct pw_resource_events resource_events = {
 	.destroy = global_unbind,
 };
 
-static void remove_pending(struct resource_data *d)
-{
-	if (d->pong_seq != 0) {
-		pw_impl_client_set_busy(pw_resource_get_client(d->resource), false);
-		d->pong_seq = 0;
-		d->impl->pending--;
-	}
-}
-
 static void impl_resource_destroy(void *data)
 {
 	struct resource_data *d = data;

